source/common/filter/tcp_proxy.cc

```diff
@@ -163,6 +163,14 @@ void TcpProxy::DownstreamCallbacks::onAboveWriteBufferHighWatermark() {
 void TcpProxy::DownstreamCallbacks::onBelowWriteBufferLowWatermark() {
   ASSERT(on_high_watermark_called_);
   on_high_watermark_called_ = false;
+
+  if (parent_.upstream_connection_->state() != Network::Connection::State::Open) {
+    // Because we flush write downstream, we can have a case where upstream has already disconnected
+    // and we are waiting to flush. If we had a watermark event during this time we should no
+    // longer touch the upstream connection.
+    return;
```

Comment:
I'm unclear why we need to do this check here. If parent_ and parent_upstream_connection_ are accessible enough to check the state, why not put this check in readDisableUpstream or even in readDisable?  The closer to the source we can put it the more likely we won't have a similar crash when someone calls these functions from another location.

Reply:
The reason I didn't do it in `readDisableUpstream` is I wasn't thinking about it that much. Can definitely move it there. The reason I didn't do it in `readDisable` is it feels to me like calling connection methods after calling close() is a programming error. I could go either way on this. If you feel it would be better to put the guard there I can move it and do a connection_impl test.

Reply:
That's reasonable, but if it's a programmer error which we've encountered and will cause a crash it seems worth something.  Maybe move to reaDDisableUpstream, and stick in ASSERT in readDisable to make it clear?

Reply:
Sure will do.